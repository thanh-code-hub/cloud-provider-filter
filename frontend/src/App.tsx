import { useCallback } from 'react';
import './App.css';
import { CloudProvider, Location } from './types/types';
import LocationSearchForm from './components/locationSearchForm';
import useFetchClouds from './hooks/useFetchClouds';

function App() {
  const { cloudProviders, isLoading, error, fetchClouds } = useFetchClouds();

  const handleLocationFound = (provider?: string, location?: Location) => {
    fetchClouds(provider, location).catch((error) => {
      console.error('Error filtering clouds provider:', error);
    });
  };

  const contentRenderer = useCallback(() => {
    if (isLoading) {
      return (
        <tr>
          <td colSpan={4} style={{ textAlign: 'center' }} role="provider-table-loading-row">
            Loading cloud data...
          </td>
        </tr>
      );
    }
    if (error) {
      return (
        <tr>
          <td colSpan={4} style={{ textAlign: 'center' }} role="provider-table-error-row">
            Failed to fetch cloud providers
          </td>
        </tr>
      );
    }
    if (cloudProviders.length) {
      return cloudProviders.map((cloudProvider: CloudProvider) => (
        <tr key={cloudProvider.cloud_name} role="provider-table-row">
          <td>{cloudProvider.cloud_name}</td>
          <td>{cloudProvider.cloud_description}</td>
          <td>{cloudProvider.provider}</td>
          <td>{cloudProvider.provider_description}</td>
        </tr>
      ));
    }
    return (
      <tr>
        <td colSpan={4} style={{ textAlign: 'center' }} role="provider-table-empty-row">
          No cloud providers found
        </td>
      </tr>
    );
  }, [isLoading, cloudProviders, error]);

  return (
    <div role="root-container">
      <LocationSearchForm onLocationFound={handleLocationFound} disabled={isLoading} />
      <table role="provider-table">
        <thead>
          <tr>
            <th>Cloud Name</th>
            <th>Cloud Description</th>
            <th>Provider</th>
            <th>Provider Description</th>
          </tr>
        </thead>
        <tbody role="provider-table-body">{contentRenderer()}</tbody>
      </table>
    </div>
  );
}

export default App;
