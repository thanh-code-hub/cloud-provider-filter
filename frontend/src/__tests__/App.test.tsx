import React from 'react';
import { act } from 'react-dom/test-utils';

import { render, screen, waitFor } from '@testing-library/react';
import App from '../App';

describe('Testing App component', () => {
  afterEach(() => {
    jest.restoreAllMocks();
  });

  test('renders correctly', () => {
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });

  test('renders failed response', async () => {
    jest.spyOn(global, 'fetch').mockImplementation(
      jest.fn(() =>
        Promise.resolve({
          ok: false,
          status: 500,
          json: () =>
            Promise.resolve({
              detail: 'Testing: some error message',
            }),
        }),
      ) as jest.Mock,
    );

    act(() => {
      render(<App />);
    });
    await waitFor(() => {
      expect(screen.getByRole('provider-table-error-row')).toBeInTheDocument();
      expect(screen.getByText('Failed to fetch cloud providers')).toBeInTheDocument();
    });
  });

  test('renders successful response', async () => {
    jest.spyOn(global, 'fetch').mockImplementation(
      jest.fn(() =>
        Promise.resolve({
          ok: true,
          status: 200,
          json: () =>
            Promise.resolve([
              {
                cloud_description: 'Africa, South Africa - Amazon Web Services: Cape Town',
                cloud_name: 'aws-af-south-1',
                geo_latitude: -3.92,
                geo_longitude: 1.42,
                provider: 'aws',
                provider_description: 'Amazon Web Services',
              },
              {
                cloud_description: 'Africa, South Africa - Azure: South Africa North',
                cloud_name: 'azure-south-africa-north',
                geo_latitude: -25.734,
                geo_longitude: 21.21,
                provider: 'azure',
                provider_description: 'Microsoft Azure',
              },
              {
                cloud_description: 'Africa, South Africa - Google Cloud: Johannesburg',
                cloud_name: 'google-africa-south1',
                geo_latitude: -16.1215,
                geo_longitude: 22.0229,
                provider: 'google',
                provider_description: 'Google Cloud Platform',
              },
            ]),
        }),
      ) as jest.Mock,
    );

    act(() => {
      render(<App />);
    });

    await waitFor(() => {
      expect(screen.getByRole('provider-table-body')).toBeInTheDocument();
      expect(screen.getAllByRole('provider-table-row')).toHaveLength(3);
    });
  });
});
