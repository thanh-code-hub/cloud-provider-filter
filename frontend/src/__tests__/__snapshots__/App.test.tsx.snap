// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing App component renders correctly 1`] = `
<div>
  <div
    role="root-container"
  >
    <div>
      <form
        role="filter-form"
      >
        <fieldset
          disabled=""
          style="padding: 0px; margin: 0px;"
        >
          <input
            placeholder="Enter provider"
            role="provider-search-input"
            type="text"
            value=""
          />
          <input
            placeholder="Enter location or address"
            role="location-search-input"
            type="text"
            value=""
          />
          <button
            role="search-button"
            type="submit"
          >
            Search cloud provider
          </button>
          <button
            role="clear-button"
          >
            Clear
          </button>
        </fieldset>
      </form>
    </div>
    <table
      role="provider-table"
    >
      <thead>
        <tr>
          <th>
            Cloud Name
          </th>
          <th>
            Cloud Description
          </th>
          <th>
            Provider
          </th>
          <th>
            Provider Description
          </th>
        </tr>
      </thead>
      <tbody
        role="provider-table-body"
      >
        <tr>
          <td
            colspan="4"
            role="provider-table-loading-row"
            style="text-align: center;"
          >
            Loading cloud data...
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
